File: maths/deltacalc.m
function [alpha,beta,gamma] = deltacalc(X,Y,Z,inverted)
%function deltacalc()
    c = 130;
    a = 70;
    b = 150;
    C1 = [c*sqrt(3)/2*cos(-150/180*pi); c*sqrt(3)/2*sin(-150/180*pi); 0];
    C2 = [c*sqrt(3)/2*cos(90/180*pi); c*sqrt(3)/2*sin(90/180*pi); 0];
    C3 = [c*sqrt(3)/2*cos(-30/180*pi); c*sqrt(3)/2*sin(-30/180*pi); 0];
    
    orthoC1 = [sin(-150/180*pi); -cos(-150/180*pi); 0];
    orthoC2 = [sin(90/180*pi); -cos(90/180*pi); 0];
    orthoC3 = [sin(-30/180*pi); -cos(-30/180*pi); 0];
    
%     X = 0;
%     Y = 0;
%     Z = 100;
    
    d = 145;
    D1 = [X+d*sqrt(3)/2*cos(-150/180*pi); Y+d*sqrt(3)/2*sin(-150/180*pi); Z];
    D2 = [X+d*sqrt(3)/2*cos(90/180*pi); Y+d*sqrt(3)/2*sin(90/180*pi); Z];
    D3 = [X+d*sqrt(3)/2*cos(-30/180*pi); Y+d*sqrt(3)/2*sin(-30/180*pi); Z];

    %alphav
    Dp1 = line_projection([0;0;Z], C1 + [0;0;Z], D1);
    d0 = norm(Dp1 - C1);
    bp = sqrt(b^2-norm(D1-Dp1)^2);

    Dp1u = (C1 - Dp1)/norm(C1 - Dp1);
    Dp1ut = cross(Dp1u, orthoC1);
    if (inverted)
        Dp1ut = - Dp1ut;
    end;
    cc_a = (bp^2 - a^2 + norm(Dp1 - C1)^2) / (2*norm(Dp1 - C1));
    cc_h = sqrt(bp^2 - cc_a^2);
    P1 = Dp1 + cc_a * Dp1u - cc_h * Dp1ut;
    
    %alpha = acos((norm(C1)^2 + norm(C1-P1)^2 - norm(P1)^2) / (2*norm(C1)*norm(C1-P1)))
    %alpha = pi - atan(P1(3)/norm([P1(1); P1(2); 0]-C1))
    if (inverted)
        alpha = asin(P1(3) / a);
    else
        alpha = pi - asin(P1(3) / a);
    end;
    
    %beta
    Dp2 = line_projection([0;0;Z], C2 + [0;0;Z], D2);
    d0 = norm(Dp2 - C2);
    bp = sqrt(b^2-norm(D2-Dp2)^2);
    
    Dp2u = (C2 - Dp2)/norm(C2 - Dp2);
    Dp2ut = -cross(Dp2u, orthoC2);
    if (inverted)
        Dp2ut = - Dp2ut;
    end
    cc_a = (bp^2 - a^2 + norm(Dp2 - C2)^2) / (2*norm(Dp2 - C2));
    cc_h = sqrt(bp^2 - cc_a^2);
    P2 = Dp2 + cc_a * Dp2u - cc_h * Dp2ut;
    
    %beta = acos((norm(C2)^2 + norm(C2-P2)^2 - norm(P2)^2) / (2*norm(C2)*norm(C2-P2)))
    %beta = pi - atan(P2(3)/norm([P2(1); P2(2); 0]-C2))
    if (inverted)
        beta = asin(P2(3) / a);
    else
        beta = pi - asin(P2(3) / a);
    end;
    
    %gamma
    Dp3 = line_projection([0;0;Z], C3 + [0;0;Z], D3);
    d0 = norm(Dp3 - C3);
    bp = sqrt(b^2-norm(D3-Dp3)^2);
    
    Dp3u = (C3 - Dp3)/norm(C3 - Dp3);
    Dp3ut = -cross(Dp3u, orthoC3);
    if (inverted)
        Dp3ut = - Dp3ut;
    end
    cc_a = (bp^2 - a^2 + norm(Dp3 - C3)^2) / (2*norm(Dp3 - C3));
    cc_h = sqrt(bp^2 - cc_a^2);
    P3 = Dp3 + cc_a * Dp3u - cc_h * Dp3ut;
    
    %gamma = acos((norm(C3)^2 + norm(C3-P3)^2 - norm(P3)^2) / (2*norm(C3)*norm(C3-P3)))
    %gamma = pi - atan(P3(3)/norm([P3(1); P3(2); 0]-C3))
    if (inverted)
        gamma = asin(P3(3) / a);
    else
        gamma = pi - asin(P3(3) / a);
    end;
    
    %alpha/pi*180, beta/pi*180, gamma/pi*180
    
%     Hp1 = C1 + [cos(pi-alpha)*a * cos(-150/180*pi); cos(pi-alpha)*a * sin(-150/180*pi); a*sin(pi-alpha)];
%     norm(Hp1-D1)
%     
%     Hp2 = C2 + [cos(pi-beta)*a * cos(90/180*pi); cos(pi-beta)*a * sin(90/180*pi); a*sin(pi-beta)];
%     norm(Hp2-D2)
%     
%     Hp3 = C3 + [cos(pi-gamma)*a * cos(-30/180*pi); cos(pi-gamma)*a * sin(-30/180*pi); a*sin(pi-gamma)];
%     norm(Hp3-D3)
end

function res = line_projection(A, B, P)
    res = A + (dot(B-A, P-A) / (norm(B-A)^2)) * (B-A);
end

File: bricks/01_stellaris_delta_calc_test/delta_calc.c
#include <math.h>

#include "delta_calc.h"

float vsqrtf_stellaris(float);

inline float v3_dot(v3 a, v3 b) {
	return a.x*b.x + a.y*b.y + a.z*b.z;
}

inline float v3_dot_p(v3 *a, v3 *b) {
	return a->x*b->x + a->y*b->y + a->z*b->z;
}

inline float v3_norm(v3 a) {
	return vsqrtf_stellaris(a.x*a.x + a.y*a.y + a.z*a.z);
}

inline float v3_norm_p(v3 *a) {
	return vsqrtf_stellaris(a->x*a->x + a->y*a->y + a->z*a->z);
}

inline float v3_norm_squared(v3 a) {
	return a.x*a.x + a.y*a.y + a.z*a.z;
}

inline float v3_norm_squared_p(v3 *a) {
	return a->x*a->x + a->y*a->y + a->z*a->z;
}

inline v3 v3_sum(v3 a, v3 b) {
	v3 res;
	res.x = a.x + b.x;
	res.y = a.y + b.y;
	res.z = a.z + b.z;
	return res;
}

inline void v3_sum_p(v3 *a, v3 *b, v3 *out) {
	out->x = a->x + b->x;
	out->y = a->y + b->y;
	out->z = a->z + b->z;
}

inline v3 v3_diff(v3 a, v3 b) {
	v3 res;
	res.x = a.x - b.x;
	res.y = a.y - b.y;
	res.z = a.z - b.z;
	return res;
}

inline void v3_diff_p(v3 *a, v3 *b, v3 *out) {
	out->x = a->x - b->x;
	out->y = a->y - b->y;
	out->z = a->z - b->z;
}

inline v3 v3_mult(v3 a, float m) {
	v3 res;
	res.x = a.x * m;
	res.y = a.y * m;
	res.z = a.z * m;
	return res;
}

inline void v3_mult_p(v3 *a, float m, v3 *out) {
	out->x = a->x * m;
	out->y = a->y * m;
	out->z = a->z * m;
}

inline void v3_cross_p(v3 *a, v3 *b, v3 *out) {
	out->x = a->y * b->z - a->z * b->y;
	out->y = a->z * b->x - a->x * b->z;
	out->z = a->x * b->y - a->y * b->x;
}

inline void line_projection3d_p(v3 *a, v3 *b, v3 *p, v3 *out) {
	v3 BmA, PmA, res;
	float coef, normed, doted;
	v3_diff_p(b, a, &BmA);
	v3_diff_p(p, a, &PmA);
	doted = v3_dot_p(&BmA, &PmA);
	normed = v3_norm_p(&BmA);
	normed = normed * normed;
	coef = doted / normed;
	v3_mult_p(&BmA, coef, &res);
	v3_sum_p(a, &res, out);
}

int delta_calc(delta_params params, float X, float Y, float Z, int inverted, float *alpha, float *beta, float *gamma) {
	//
	// alpha
	//
	v3 C1, D1, orthoC1, Dp1;
	C1.x = params.c * -0.750000000000000; //sqrt(3)/2. * cos(-150./180.*M_PI);
	C1.y = params.c * -0.433012701892219; //sqrt(3)/2. * sin(-150./180.*M_PI);
	C1.z = 0;

	D1.x = X + (params.d * -0.750000000000000);
	D1.y = Y + (params.d * -0.433012701892219);
	D1.z = Z;

	orthoC1.x = -0.500000000000000;
	orthoC1.y = 0.866025403784439;
	orthoC1.z = 0;

	v3 lp1, lp2, D1mDp1;
	lp1.x = 0; lp1.y = 0; lp1.z = Z;
	float bp;
	v3_sum_p(&lp1, &C1, &lp2);
	line_projection3d_p(&lp1, &lp2, &D1, &Dp1);
	v3_diff_p(&D1, &Dp1, &D1mDp1);
	float D1Dp1norm = v3_norm_squared_p(&D1mDp1);
	bp = vsqrtf_stellaris(params.b*params.b - D1Dp1norm);

	v3 Dp1u, Dp1ut, C1mDp1;
	float C1mDp1_norm;
	v3_diff_p(&C1, &Dp1, &C1mDp1);
	C1mDp1_norm = v3_norm_p(&C1mDp1);
	v3_mult_p(&C1mDp1, 1/C1mDp1_norm, &Dp1u);
	v3_cross_p(&Dp1u, &orthoC1, &Dp1ut);
	if (inverted) {
		v3_mult_p(&Dp1ut, -1, &Dp1ut);
	}

	float cc_a, cc_h;
	v3 P1, buffer;
	cc_a = (bp*bp - params.a*params.a + C1mDp1_norm*C1mDp1_norm) / (2 * C1mDp1_norm);
	cc_h = vsqrtf_stellaris(bp*bp - cc_a * cc_a);
	v3_mult_p(&Dp1u, cc_a, &buffer);
	v3_sum_p(&Dp1, &buffer, &P1);
	v3_mult_p(&Dp1ut, cc_h, &buffer);
	v3_diff_p(&P1, &buffer, &P1);

	if (inverted)
		*alpha = asin(P1.z / params.a);
	else
		*alpha = M_PI - asin(P1.z / params.a);

	//
	// beta
	//
	v3 C2, D2, orthoC2, Dp2;
	C2.x = params.c * 0; //sqrt(3)/2. * cos(90./180.*M_PI);
	C2.y = params.c * 0.866025403784439; //sqrt(3)/2. * sin(90./180.*M_PI);
	C2.z = 0;

	D2.x = X + (params.d * 0);
	D2.y = Y + (params.d * 0.866025403784439);
	D2.z = Z;

	orthoC2.x = 1;
	orthoC2.y = 0;
	orthoC2.z = 0;

	v3 D2mDp2;
	lp1.x = 0; lp1.y = 0; lp1.z = Z;
	v3_sum_p(&lp1, &C2, &lp2);
	line_projection3d_p(&lp1, &lp2, &D2, &Dp2);
	v3_diff_p(&D2, &Dp2, &D2mDp2);
	float D2Dp2norm = v3_norm_squared_p(&D2mDp2);
	bp = vsqrtf_stellaris(params.b*params.b - D2Dp2norm);

	v3 Dp2u, Dp2ut, C2mDp2;
	float C2mDp2_norm;
	v3_diff_p(&C2, &Dp2, &C2mDp2);
	C2mDp2_norm = v3_norm_p(&C2mDp2);
	v3_mult_p(&C2mDp2, 1/C2mDp2_norm, &Dp2u);
	v3_cross_p(&Dp2u, &orthoC2, &Dp2ut);
	if (inverted) {
		v3_mult_p(&Dp2ut, -1, &Dp2ut);
	}

	v3 P2;
	cc_a = (bp*bp - params.a*params.a + C2mDp2_norm*C2mDp2_norm) / (2 * C2mDp2_norm);
	cc_h = vsqrtf_stellaris(bp*bp - cc_a * cc_a);
	v3_mult_p(&Dp2u, cc_a, &buffer);
	v3_sum_p(&Dp2, &buffer, &P2);
	v3_mult_p(&Dp2ut, cc_h, &buffer);
	v3_diff_p(&P2, &buffer, &P2);

	if (inverted)
		*beta = asin(P2.z / params.a);
	else
		*beta = M_PI - asin(P2.z / params.a);

	//
	// gamma
	//
	v3 C3, D3, orthoC3, Dp3;
	C3.x = params.c * 0.750000000000000; //sqrt(3)/2. * cos(-30./180.*M_PI);
	C3.y = params.c * -0.433012701892219; //sqrt(3)/2. * sin(-30./180.*M_PI);
	C3.z = 0;

	D3.x = X + (params.d * 0.750000000000000);
	D3.y = Y + (params.d * -0.433012701892219);
	D3.z = Z;

	orthoC3.x = -0.5;
	orthoC3.y = -0.866025403784439;
	orthoC3.z = 0;

	v3 D3mDp3;
	lp1.x = 0; lp1.y = 0; lp1.z = Z;
	v3_sum_p(&lp1, &C3, &lp2);
	line_projection3d_p(&lp1, &lp2, &D3, &Dp3);
	v3_diff_p(&D3, &Dp3, &D3mDp3);
	float D3Dp3norm = v3_norm_squared_p(&D3mDp3);
	bp = vsqrtf_stellaris(params.b*params.b - D3Dp3norm);

	v3 Dp3u, Dp3ut, C3mDp3;
	float C3mDp3_norm;
	v3_diff_p(&C3, &Dp3, &C3mDp3);
	C3mDp3_norm = v3_norm_p(&C3mDp3);
	v3_mult_p(&C3mDp3, 1/C3mDp3_norm, &Dp3u);
	v3_cross_p(&Dp3u, &orthoC3, &Dp3ut);
	if (inverted) {
		v3_mult_p(&Dp3ut, -1, &Dp3ut);
	}

	v3 P3;
	cc_a = (bp*bp - params.a*params.a + C3mDp3_norm*C3mDp3_norm) / (2 * C3mDp3_norm);
	cc_h = vsqrtf_stellaris(bp*bp - cc_a * cc_a);
	v3_mult_p(&Dp3u, cc_a, &buffer);
	v3_sum_p(&Dp3, &buffer, &P3);
	v3_mult_p(&Dp3ut, cc_h, &buffer);
	v3_diff_p(&P3, &buffer, &P3);

	if (inverted)
		*gamma = asin(P3.z / params.a);
	else
		*gamma = M_PI - asin(P3.z / params.a);

	return 0;
}
